gradoop-data-integration/src/main/java/org/gradoop/dataintegration/transformation/VertexDeduplication.java

```diff
@@ -0,0 +1,113 @@
+import org.gradoop.dataintegration.transformation.functions.GetPropertiesAsList;
+import org.gradoop.dataintegration.transformation.functions.MarkDuplicatesInGroup;
+import org.gradoop.dataintegration.transformation.functions.RemoveMarkedDuplicates;
+import org.gradoop.dataintegration.transformation.functions.UpdateEdgeSource;
+import org.gradoop.dataintegration.transformation.functions.UpdateEdgeTarget;
+import org.gradoop.flink.model.api.epgm.BaseGraph;
+import org.gradoop.flink.model.api.epgm.BaseGraphCollection;
+import org.gradoop.flink.model.api.operators.UnaryBaseGraphToBaseGraphOperator;
+import org.gradoop.flink.model.impl.functions.epgm.ByLabel;
+import org.gradoop.flink.model.impl.functions.epgm.SourceId;
+import org.gradoop.flink.model.impl.functions.epgm.TargetId;
+import org.gradoop.flink.model.impl.functions.filters.Not;
+
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Deduplicates vertices based on some attribute.<p>
+ * Given a label and a list of property keys, this operator will find all vertices of that label and condense
+ * duplicates to one vertex. Two (or more) vertices are considered duplicates if all property values,
+ * respective to the list of given keys, are equal.<p>
+ * Other attributes of the vertices are ignored, the new (condensed) vertex will have the attributes of some
+ * (random) vertex of its duplicates.
+ * <p>
+ * This will create no new elements and retain the graph head.
+ *
+ * @param <G>  The graph head type.
+ * @param <V>  The vertex type.
+ * @param <E>  The edge type.
+ * @param <LG> The type of the graph.
+ * @param <GC> The type of the graph collection.
+ */
+public class VertexDeduplication<
+  G extends GraphHead,
+  V extends Vertex,
+  E extends Edge,
+  LG extends BaseGraph<G, V, E, LG, GC>,
+  GC extends BaseGraphCollection<G, V, E, LG, GC>>
+  implements UnaryBaseGraphToBaseGraphOperator<LG> {
+
+  /**
+   * The label of vertices to deduplicate.
+   */
+  private final String label;
+
+  /**
+   * The list of property keys used to determine equal vertices.
+   */
+  private final List<String> propertyKeys;
+
+  /**
+   * Creates a new instance of this vertex deduplication operator.
+   *
+   * @param label        The label of vertices to deduplicate.
+   * @param propertyKeys The properties used to check if two (or more) vertices of that label are duplicates.
+   */
+  public VertexDeduplication(String label, List<String> propertyKeys) {
+    this.label = Objects.requireNonNull(label);
+    this.propertyKeys = Objects.requireNonNull(propertyKeys);
+  }
+
+  @Override
+  public LG execute(LG graph) {
+    DataSet<V> annotatedVertices = graph.getVerticesByLabel(label)
+      .groupBy(new GetPropertiesAsList<>(propertyKeys))
+      .reduceGroup(new MarkDuplicatesInGroup<>())
+      .name("Mark duplicate vertices");
+    DataSet<Tuple2<GradoopId, GradoopId>> vertexToDedupVertex = annotatedVertices
+      .flatMap(new CreateMappingFromMarkedDuplicates<>())
```

Comment:
I think you can save several flink operators by creating the mapping in the `reduceGroup` step and writing it as map property on each original element. Then you can remove the duplicates directly in the `reduceGroup` as well. The data flow should be similar, but you reduce the overhead.

Reply:
I think this would destroy scalablity because in the worst case all vertices are in one group, leading to one element storing all other ids.
Storing the mapping on vertices should be good for many small groups, but I don't think that we should assume this